stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -2551,42 +2587,67 @@ final internal class _Native${Self}Storage<${TypeParameters}> :
   internal static func create(capacity: Int) -> Storage {
     let numWordsForBitmap = _UnsafeBitMap.sizeInWords(forSizeInBits: capacity)
 %if Self == 'Dictionary':
-    let storage = Builtin.allocWithTailElems_3(Storage.self,
+    let buffer = Builtin.allocWithTailElems_3(TypedBuffer.self,
         numWordsForBitmap._builtinWordValue, UInt.self,
         capacity._builtinWordValue, Key.self,
         capacity._builtinWordValue, Value.self)
 %else:
-    let storage = Builtin.allocWithTailElems_2(Storage.self,
+    let buffer = Builtin.allocWithTailElems_2(TypedBuffer.self,
         numWordsForBitmap._builtinWordValue, UInt.self,
         capacity._builtinWordValue, Key.self)
 %end
     
     // We can initialize by assignment because _HashedContainerStorageHeader
     // is a trivial type, i.e. contains no references.
-    storage._body = StorageHeader(capacity: capacity)
- 
+    buffer._body = StorageHeader(capacity: capacity)
+    
+    var storage = Storage(_buffer: buffer)
+
     let initializedEntries = _UnsafeBitMap(
         storage: storage._initializedHashtableEntriesBitMapStorage,
         bitCount: capacity)
     initializedEntries.initializeToZero()
 
     // Initialize header
     storage._body.initializedEntries = initializedEntries
-    storage._body.keys = storage._keys
+    storage._body.keys = UnsafeMutableRawPointer(storage._keys)
 %if Self == 'Dictionary':
-    storage._body.values = storage._values
+    storage._body.values = UnsafeMutableRawPointer(storage._values)
 %end
 
     return storage
   }
 
+  init(_buffer: RawBuffer) {
+    self._buffer = _buffer
+  }
+
+  init() {
+    self._buffer = RawBuffer.empty
+  }
+}
+
+
+/// An instance of this class has all `${Self}` data tail-allocated.
+/// Enough bytes are allocated to hold the bitmap for marking valid entries,
+/// keys, and values. The data layout starts with the bitmap, followed by the
+/// keys, followed by the values.
+@_versioned
+final internal class _Native${Self}Buffer<${TypeParameters}> :
+  _RawNative${Self}Buffer {
+
+%if Self == 'Set': # Set needs these to keep signatures simple.
+  internal typealias Key = ${TypeParameters}
+  internal typealias Value = ${TypeParameters}
+%end
+
   deinit {
-    let capacity = _capacity
+    let capacity = _body.capacity
     let initializedEntries = _UnsafeBitMap(
         storage: _initializedHashtableEntriesBitMapStorage, bitCount: capacity)
-    let keys = _keys
+    let keys = _body.keys.assumingMemoryBound(to: Key.self)
```

Comment:
_body doesn't know the right type, because it's on RawBuffer. Had to do weird things in here because I wasn't sure what the implication of wrapping self in a NativeStorage during deinit would be.